openpype/plugins/publish/integrate_inputlinks.py

```diff
@@ -0,0 +1,132 @@
+            - link workfile as input of all publishing instances
+        * else:
+            - show "no workfile" warning
+        * link instances' inputs if it's data has "inputVersions" entry
+        * Write into database
+
+        inputVersions:
+            The "inputVersions" in instance.data should be a list of
+            version document's Id (str or ObjectId), which are the
+            dependencies of the publishing instance that should be
+            extracted from working scene by the DCC specific publish
+            plugin.
+
+        """
+        workfile = None
+        publishing = []
+
+        for instance in context:
+            if not instance.data.get("publish", True):
+                # Skip inactive instances
+                continue
+
+            version_doc = instance.data.get("versionEntity")
+            if not version_doc:
+                self.log.debug("Instance %s doesn't have version." % instance)
+                continue
+
+            version_data = version_doc.get("data", {})
+            families = version_data.get("families", [])
+
+            if "workfile" in families:
+                workfile = instance
+            else:
+                publishing.append(instance)
+
+        if workfile is None:
+            self.log.warn("No workfile in this publish session.")
+        else:
+            workfile_version_doc = workfile.data["versionEntity"]
+            # link all loaded versions in scene into workfile
+            for version in context.data.get("loadedVersions", []):
+                self.add_link(
+                    link_type="reference",
+                    input_id=version["version"],
+                    version_doc=workfile_version_doc,
+                )
+            # link workfile to all publishing versions
+            for instance in publishing:
+                self.add_link(
+                    link_type="generative",
+                    input_id=workfile_version_doc["_id"],
+                    version_doc=instance.data["versionEntity"],
+                )
+
+        # link versions as dependencies to the instance
+        for instance in publishing:
+            for input_version in instance.data.get("inputVersions") or []:
+                self.add_link(
+                    link_type="generative",
+                    input_id=input_version,
+                    version_doc=instance.data["versionEntity"],
+                )
+
+        publishing.append(workfile)
+        self.write_links_to_database(publishing)
+
+    def add_link(self, link_type, input_id, version_doc):
+        """Add dependency link data into version document
+
+        Args:
+            link_type (str): Type of link, one of 'reference' or 'generative'
+            input_id (str or ObjectId): Document Id of input version
+            version_doc (dict): The version document that takes the input
+
+        Returns:
+            None
+
+        """
+        from collections import OrderedDict
```

Comment:
I think you can put these imports to the top of file

Reply:
This is kind of like a quirks of mine when I writing pyblish plugins, always afraid of modules getting garbage collected when previous plugin that also used the same modules has been processed.

I'll move them to the top and give it a go.